fem/bilininteg_hcurl.cpp

```diff
@@ -0,0 +1,2059 @@
+   const int dims = el->GetDim();
+   MFEM_VERIFY(dims == 2 || dims == 3, "");
+
+   const int symmDims = (dims * (dims + 1)) / 2; // 1x1: 1, 2x2: 3, 3x3: 6
+   const int nq = ir->GetNPoints();
+   dim = mesh->Dimension();
+   MFEM_VERIFY(dim == 2 || dim == 3, "");
+
+   ne = fes.GetNE();
+   geom = mesh->GetGeometricFactors(*ir, GeometricFactors::JACOBIANS);
+   mapsC = &el->GetDofToQuad(*ir, DofToQuad::TENSOR);
+   mapsO = &el->GetDofToQuadOpen(*ir, DofToQuad::TENSOR);
+   dofs1D = mapsC->ndof;
+   quad1D = mapsC->nqpt;
+
+   MFEM_VERIFY(dofs1D == mapsO->ndof + 1 && quad1D == mapsO->nqpt, "");
+
+   pa_data.SetSize(symmDims * nq * ne, Device::GetMemoryType());
+
+   Vector coeff(ne * nq);
+   coeff = 1.0;
+   if (Q)
+   {
+      for (int e=0; e<ne; ++e)
+      {
+         ElementTransformation *tr = mesh->GetElementTransformation(e);
+         for (int p=0; p<nq; ++p)
+         {
+	   coeff[p + (e * nq)] = Q->Eval(*tr, ir->IntPoint(p));
+         }
+      }
+   }
+
+   if (el->GetDerivType() == mfem::FiniteElement::CURL && dim == 3)
+   {
+      PAHcurlSetup3D(quad1D, ne, ir->GetWeights(), geom->J,
+                     coeff, pa_data);
+   }
+   else if (el->GetDerivType() == mfem::FiniteElement::CURL && dim == 2)
+   {
+      PAHcurlSetup2D(quad1D, ne, ir->GetWeights(), geom->J,
+                     coeff, pa_data);
+   }
+   else
+   {
+      MFEM_ABORT("Unknown kernel.");
+   }
+
+   dof_map = el->GetDofMap();
+}
+
+static void PAHcurlMassApply2D(const int D1D,
+                               const int Q1D,
+                               const int NE,
+                               const Array<int> &_dof_map,
+                               const Array<double> &_Bo,
+                               const Array<double> &_Bc,
+                               const Array<double> &_Bot,
+                               const Array<double> &_Bct,
+                               const Vector &_op,
+                               const Vector &_x,
+                               Vector &_y)
+{
+   constexpr static int VDIM = 2;
+
+   auto dof_map = Reshape(_dof_map.Read(), 2*(D1D-1)*D1D);
+   auto Bo = Reshape(_Bo.Read(), Q1D, D1D-1);
+   auto Bc = Reshape(_Bc.Read(), Q1D, D1D);
+   auto Bot = Reshape(_Bot.Read(), D1D-1, Q1D);
+   auto Bct = Reshape(_Bct.Read(), D1D, Q1D);
+   auto op = Reshape(_op.Read(), Q1D*Q1D, 3, NE);
+   auto x = Reshape(_x.Read(), 2*(D1D-1)*D1D, NE);
+   auto y = Reshape(_y.ReadWrite(), 2*(D1D-1)*D1D, NE);
+
+   MFEM_FORALL(e, NE,
+   {
+      double mass[MAX_Q1D][MAX_Q1D][VDIM];
+
+      for (int qy = 0; qy < Q1D; ++qy)
+      {
+         for (int qx = 0; qx < Q1D; ++qx)
+         {
+            for (int c = 0; c < VDIM; ++c)
+            {
+               mass[qy][qx][c] = 0.0;
+            }
+         }
+      }
+
+      int osc = 0;
+
+      for (int c = 0; c < VDIM; ++c)  // loop over x, y components
+      {
+         const int D1Dy = (c == 1) ? D1D - 1 : D1D;
+         const int D1Dx = (c == 0) ? D1D - 1 : D1D;
```

Comment:
Would it perform better if I used int[] arrays, e.g. D1Dx[0] = D1D-1; D1Dx[1] = D1D; I don't have experience with the performance impact of such optimizations, so please let me know if you think this would help.

Reply:
I think we'll postpone optimization for another PR, the feature is too important to stop it because of this.